File: leetcode/editor/cn/219_存在重复元素 II_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when219_thenSuccess(t *testing.T) {
	assert.Equal(t, 1, 1, "shouldEqual")
}

//leetcode submit region begin(Prohibit modification and deletion)
func containsNearbyDuplicate(nums []int, k int) bool {
	ans := false
	// map used as a Set
	window := map[int]struct{}{}
	for i, item := range nums {
		// check item in Set
		if _, exists := window[item]; exists {
			ans = true
			break
		}
		// add to Set
		window[item] = struct{}{}
		if len(window) > k {
			// remove from Set
			delete(window, nums[i-k])
		}
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/219_存在重复元素 II_test.py
import unittest
from typing import List


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        duplicate = self.solution.containsNearbyDuplicate([1, 2, 3, 1], 3)
        self.assertTrue(duplicate)

    def test_givenNormalCase2_when_thenSuccess(self):
        duplicate = self.solution.containsNearbyDuplicate([1, 2, 3, 4, 1], 3)
        self.assertFalse(duplicate)


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        # 滑动窗口
        ans = False
        window: set[int] = set()
        for idx, item in enumerate(nums):
            if item in window:
                ans = True
                break
            else:
                window.add(item)
            if len(window) > k:
                window.remove(nums[idx - k])
        return ans
        pass


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
